src/prefect/deployments/steps/core.py

```diff
@@ -191,10 +203,97 @@ async def run_steps(
+                serialized_steps,
+                failed_step=serialized_steps[-1] if serialized_steps else None,
+                deployment=deployment,
+            )
             raise StepExecutionError(f"Encountered error while running {fqn}") from exc
+    _emit_pull_steps_event(serialized_steps, failed_step=None, deployment=deployment)
     return upstream_outputs
 
 
-def _get_step_fully_qualified_name_and_inputs(step: Dict) -> Tuple[str, Dict]:
+def _get_step_fully_qualified_name_and_inputs(step: dict) -> tuple[str, dict]:
     step = deepcopy(step)
     return step.popitem()
+
+
+def _serialize_step_for_event(
+    index: int, qualified_name: str, inputs: dict[str, Any]
+) -> dict[str, Any]:
+    """
+    Serialize a step's metadata for the pull-steps.executed event payload.
+
+    IMPORTANT: This function must be called with PRE-TEMPLATED inputs to avoid
+    exposing secrets in event payloads. Template strings like "{{ prefect.blocks.secret.name }}"
+    are preserved, while templating (which resolves these to actual secret values)
+    happens later in run_step().
+
+    Args:
+        index: The step's index in the execution sequence
+        qualified_name: The fully qualified name of the step function
+        inputs: The raw, pre-templated inputs for the step
+
+    Returns:
+        A dictionary containing step metadata safe for event emission
+    """
+    sanitized_inputs = {
+        key: value for key, value in inputs.items() if key not in RESERVED_KEYWORDS
+    }
+    return {
+        "index": index,
+        "qualified_name": qualified_name,
+        "step_name": qualified_name.split(".")[-1],
+        "id": inputs.get("id"),
+        "inputs": sanitized_inputs,
+    }
+
+
+def _emit_pull_steps_event(
+    serialized_steps: list[dict[str, Any]],
+    *,
+    failed_step: dict[str, Any] | None,
+    deployment: Any | None = None,
+) -> None:
+    if not serialized_steps:
+        return
+
+    # Use _EnvironmentRunContext to access flow run ID from environment
+    ctx = _EnvironmentRunContext.from_environment()
+    if not ctx or not ctx.flow_run_id:
+        return
+
+    payload = {
+        "count": len(serialized_steps),
+        "steps": serialized_steps,
+    }
+    if failed_step is not None:
+        payload["failed_step"] = failed_step
+    resource = {
+        "prefect.resource.id": f"prefect.flow-run.{ctx.flow_run_id}",
+    }
+
+    # Build related resources
+    related: list[dict[str, str]] = []
+    if deployment:
+        # Add deployment as a related resource
+        related.append(
+            {
+                "prefect.resource.id": f"prefect.deployment.{deployment.id}",
+                "prefect.resource.role": "deployment",
+            }
+        )
+
+    try:
+        emit_event(
+            event="prefect.flow-run.pull-steps.executed",
+            resource=resource,
+            related=related if related else None,
+            payload=payload,
+        )
```

Comment:
We've had issues with using `emit_event` outside of flow runs before (with heartbeat events in the `Runner`), so it might be better to directly use the client and force it to send events without buffering.